### PR TITLE
Debounce component search input

### PR DESCRIPTION
**Pull request by Mbeaulne**
## Description

The component search input now debounces user keystrokes by 200ms before updating the query state used for lexical search and AI search. A new `DebouncedComponentSearchInput` component manages its own local input value and only commits to the parent after the debounce delay, preventing expensive lexical search operations from running on every keystroke. `useDeferredValue` is applied to the committed query so that React can deprioritize the search re-renders, keeping the input responsive. The AI search button is now enabled only after the debounced query has settled, and the test has been updated to wait for that enabled state before clicking.

## Type of Change

- [ ] Bug fix
- [ ] New feature
- [x] Improvement
- [ ] Cleanup/Refactor
- [ ] Breaking change
- [ ] Documentation update

## Checklist

- [ ] I have tested this does not break current pipelines / runs functionality
- [ ] I have tested the changes on staging

## Test Instructions

1. Open the Dashboard Components view.
2. Type rapidly into the search box and confirm the UI remains responsive without lag on each keystroke.
3. Pause typing and confirm search results appear after ~200ms.
4. Verify the AI search button only becomes enabled once the debounced query has committed.
5. Click AI search and confirm it runs against the expected candidate pool.

## Additional Comments

The `onCommit` callback reference is stabilised via a `useRef` so the debounce `useEffect` does not restart its timer when the parent re-renders with a new callback identity.

### CODE REVIEW

**Comment by Mbeaulne** [comment]
* **#2462** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2462?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2461** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2461?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2459** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2459?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2458** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2458?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2457** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2457?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2456** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2456?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2454** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2454?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2453** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2453?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2452** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2452?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2451** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2451?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2450** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2450?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2449** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2449?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2448** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2448?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2447** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2447?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2444** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2444?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2433** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2433?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2433?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#2432** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2432?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2431** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2431?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2430** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2430?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2429** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2429?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2428** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2428?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2427** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2427?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2426** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2426?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2425** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2425?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2424** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2424?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2423** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2423?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by camielvs** [comment]
## 🤖 Code review — Debounce component search input

Small, sensible perf fix for the top of the stack — by #2432 every keystroke could trigger IDF scoring (#2426), fuzzy Levenshtein (#2427), and embedding work, so gating input is well-justified. The `DebouncedComponentSearchInput` implementation is clean: the `onCommitRef` pattern keeps the latest callback without resetting the 200ms timer, the cleanup is correct, the `aria-label` is preserved, and updating the test to `await` the button becoming enabled is the right adjustment.

Findings:

- **Debounce + `useDeferredValue` together looks redundant.** This PR introduces *both*: a 200ms debounce inside the new input component **and** `useDeferredValue(query)` in the parent. But the input is now a decoupled child with its own local state, so the parent's `query` only changes every 200ms and there's no longer a fast-updating controlled input for `useDeferredValue` to keep responsive. The net effect of the deferred layer here is mostly an extra frame of staleness on top of the debounce. I'd keep one mechanism — the standalone debounced input alone is the simpler choice. If both are deliberate, a one-line comment on what each buys would prevent a future "why two?".

- **Only the Dashboard is debounced; the Editor isn't.** `ComponentSearchV2Content` still uses immediate `onChange` + `useDeferredValue`. Given the branch name (`fix_search_input_lag`) and that the same heavy per-keystroke search runs in both surfaces, was leaving the Editor out intentional? If the lag affects both, the Editor would benefit from the same treatment (and it'd be a natural place to share the `DebouncedComponentSearchInput` component rather than duplicating later).

- **Minor — input is no longer parent-controllable.** Moving the value into child local state means the parent can't programmatically set/clear the query anymore. Nothing does that today, so it's fine, but worth noting if a "clear search" affordance is ever added. Also an initial empty `onCommit("")` fires ~200ms after mount — harmless no-op.

- **Minor — slight staleness on the AI-search click path.** `trimmedQuery` now derives from `deferredQuery` and is read by `startAiSearch`; under load a click could momentarily act on a marginally stale query. The test's `waitFor` covers it; real-world risk is low.

Overall good to go once the debounce-vs-deferred intent is clarified.

File: src/routes/Dashboard/DashboardComponentsV2View.tsx [hunk 1]
```diff
@@ -471,6 +478,44 @@ function mergeUniqueMatches(
   return merged;
 }
 
+function DebouncedComponentSearchInput({
```
Comment (Mbeaulne) [inline src/routes/Dashboard/DashboardComponentsV2View.tsx:471]: > 🤖 This is an AI-generated code review comment.

**[MEDIUM]** `DebouncedComponentSearchInput` (with its `localValue` + `onCommitRef` + two `useEffect`s + `setTimeout` debounce) is duplicated near-verbatim here and in `ComponentSearchV2Content.tsx` — only the `<Input>` styling differs. This is new duplicated logic (echoes the `mergeUniqueMatches` dup). Consider extracting the debounce into a shared hook (e.g. `useDebouncedSearchValue(onCommit)` returning `[localValue, setLocalValue]`) and letting each surface render its own styled `<Input>`.

File: src/routes/Dashboard/DashboardComponentsV2View.tsx [hunk 2]
```diff
@@ -518,9 +563,11 @@ export const DashboardComponentsV2View = () => {
   const aiDescriptionsEnabled = useFlagValue(
     "component-search-v2-ai-descriptions",
   );
+  const deepAiSearchEnabled = useFlagValue("component-search-v2-deep-ai");
   const { backendUrl, configured, available } = useBackend();
   const { config: aiConfig } = useAiProviderSettings();
   const [query, setQuery] = useState("");
+  const deferredQuery = useDeferredValue(query);
```
Comment (Mbeaulne) [inline src/routes/Dashboard/DashboardComponentsV2View.tsx]: > 🤖 This is an AI-generated code review comment.

**[LOW]** Double deferral: the input is now independently controlled by its own `localValue`, so it stays responsive regardless of the parent render cost — which was the main thing `useDeferredValue(query)` bought. With the 200ms debounce already throttling commits, `useDeferredValue` is now largely redundant here (and in `ComponentSearchV2Content.tsx`). Consider dropping it and passing `query` directly.

File: src/routes/Dashboard/DashboardComponentsV2View.tsx [hunk 3]
```diff
@@ -471,6 +478,44 @@ function mergeUniqueMatches(
...
+  const onCommitRef = useRef(onCommit);
+
+  useEffect(() => {
+    onCommitRef.current = onCommit;
+  }, [onCommit]);
+
+  useEffect(() => {
+    const timeout = window.setTimeout(() => {
+      onCommitRef.current(localValue);
+    }, SEARCH_QUERY_DEBOUNCE_MS);
```
Comment (Mbeaulne) [inline src/routes/Dashboard/DashboardComponentsV2View.tsx]: > 🤖 This is an AI-generated code review comment.

**[LOW]** The `[localValue]` debounce effect runs on mount with `localValue === ""`, so it fires a `onCommit("")` ~200ms after mount (harmless `setQuery("")`, but it also runs `clearRerank()` on this surface). Consider skipping the commit while the value still equals the last committed value, so no no-op commit fires on mount.

File: src/routes/Dashboard/DashboardComponentsV2View.tsx (hunk 1 above)
Comment (Mbeaulne) [reply src/routes/Dashboard/DashboardComponentsV2View.tsx:471]: > 🤖 This is an AI-generated response.

Fixed (working tree, pending push) — extracted the debounce into a shared `useDebouncedSearchValue(onCommit)` hook in `src/hooks/useDebouncedSearchValue.ts` (registered in `react-compiler.config.js` alongside the other search hook). Both `DebouncedComponentSearchInput` wrappers now just call the hook and render their own styled `<Input>`, so the localValue/ref/effect/setTimeout logic exists once.

File: src/routes/Dashboard/DashboardComponentsV2View.tsx (hunk 2 above)
Comment (Mbeaulne) [reply src/routes/Dashboard/DashboardComponentsV2View.tsx]: > 🤖 This is an AI-generated response.

Fixed (working tree, pending push) — removed `useDeferredValue` from both the dashboard and `ComponentSearchV2Content`; `query` is passed directly now. The debounced input (its own `localValue`) already keeps the field responsive and throttles commits to ≤1/200ms, so the extra defer layer was redundant.

File: src/routes/Dashboard/DashboardComponentsV2View.tsx (hunk 3 above)
Comment (Mbeaulne) [reply src/routes/Dashboard/DashboardComponentsV2View.tsx]: > 🤖 This is an AI-generated response.

Fixed (working tree, pending push) — the shared hook tracks the last committed value (`lastCommittedRef`) and skips scheduling a commit while `localValue` still equals it, so the no-op `onCommit("")` on mount no longer fires. Real input still commits normally (and clearing back to "" commits, since it differs from the last committed non-empty value).

File: src/routes/v2/pages/Editor/components/ComponentSearchV2Content.tsx [hunk 4]
```diff
@@ -1,19 +1,42 @@
...
+  onCommit: (value: string) => void;
+}) {
+  const [localValue, setLocalValue] = useDebouncedSearchValue(onCommit);
+
+  return (
+    <Input
+      type="text"
+      data-testid="search-input"
+      placeholder="Search components..."
+      className="w-full pl-8 text-sm h-8 focus-visible:ring-gray-400/50"
```
Comment (maxy-shpfy) [inline src/routes/v2/pages/Editor/components/ComponentSearchV2Content.tsx:27]: > 🤖 This is an AI-generated code review comment.

The extracted debounced search input introduces a new Input primitive usage in a route component with raw Tailwind passed through className. Tangle UI standards reserve className escape hatches for local primitives; route-level UI should use semantic props …

Suggested fix: Move the sizing/positioning into a local raw input primitive/wrapper or extend the Input primitive with the needed semantic props, then render the debounced component without passing className to Input. Apply the same…

File: src/routes/Dashboard/DashboardComponentsV2View.tsx [hunk 5]
```diff
@@ -471,6 +472,28 @@ function mergeUniqueMatches(
...
+
+  return (
+    <Input
+      type="search"
+      placeholder="e.g. train_test_split, pandas, clean up my data"
+      value={localValue}
+      onChange={(event) => setLocalValue(event.target.value)}
+      aria-label="Search components"
+      disabled={disabled}
+      className="flex-1"
```
Comment (maxy-shpfy) [inline src/routes/Dashboard/DashboardComponentsV2View.tsx:488]: > 🤖 This is an AI-generated code review comment.

The new DebouncedComponentSearchInput keeps layout concerns on the Input primitive via className. The route should compose layout with Tangle layout primitives or a local raw primitive instead of styling a shared UI primitive directly.

Suggested fix: Let the surrounding InlineStack/layout wrapper own flex behavior, or add a semantic sizing prop to the local search-input component so the shared Input primitive is not passed className from the route.

**Review by maxy-shpfy — APPROVED** [review]

**Comment by Mbeaulne** [comment]
### Merge activity

* **Jun 26, 7:34 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/TangleML/tangle-ui/2433).
* **Jun 26, 8:22 PM UTC**: [Graphite](https://app.graphite.com/github/pr/TangleML/tangle-ui/2433) rebased this pull request as part of a merge.
* **Jun 26, 8:25 PM UTC**: @Mbeaulne merged this pull request with [Graphite](https://app.graphite.com/github/pr/TangleML/tangle-ui/2433).
